src/web/ng/Response.cpp

```diff
@@ -0,0 +1,116 @@
+
+#include "web/ng/Response.hpp"
+
+#include "util/Assert.hpp"
+#include "util/build/Build.hpp"
+#include "web/ng/Request.hpp"
+
+#include <boost/asio/buffer.hpp>
+#include <boost/beast/http/field.hpp>
+#include <boost/beast/http/message.hpp>
+#include <boost/beast/http/status.hpp>
+#include <boost/beast/http/string_body.hpp>
+#include <boost/json/object.hpp>
+#include <boost/json/serialize.hpp>
+#include <fmt/core.h>
+
+#include <optional>
+#include <string>
+#include <string_view>
+#include <type_traits>
+#include <utility>
+
+namespace http = boost::beast::http;
+namespace web::ng {
+
+namespace {
+
+std::string_view
+asString(Response::HttpData::ContentType type)
+{
+    switch (type) {
+        case Response::HttpData::ContentType::TextHtml:
+            return "text/html";
+        case Response::HttpData::ContentType::ApplicationJson:
+            return "application/json";
+    }
+    ASSERT(false, "Unknown content type");
+    std::unreachable();
+}
+
+template <typename MessageType>
+std::optional<Response::HttpData>
+makeHttpData(http::status status, Request const& request)
+{
+    if (request.isHttp()) {
+        auto const& httpRequest = request.asHttpRequest()->get();
+        auto constexpr contentType = std::is_same_v<std::remove_cvref_t<MessageType>, std::string>
+            ? Response::HttpData::ContentType::TextHtml
+            : Response::HttpData::ContentType::ApplicationJson;
+        return Response::HttpData{
+            .status = status,
+            .contentType = contentType,
+            .keepAlive = httpRequest.keep_alive(),
+            .version = httpRequest.version()
+        };
+    }
+    return std::nullopt;
+}
+}  // namespace
+
+Response::Response(boost::beast::http::status status, std::string message, Request const& request)
+    : message_(std::move(message)), httpData_{makeHttpData<decltype(message)>(status, request)}
+{
+}
+
+Response::Response(boost::beast::http::status status, boost::json::object const& message, Request const& request)
+    : message_(boost::json::serialize(message)), httpData_{makeHttpData<decltype(message)>(status, request)}
+{
+}
+
+std::string const&
+Response::message() const
+{
+    return message_;
+}
+
+http::response<http::string_body>
+Response::intoHttpResponse() &&
+{
+    ASSERT(httpData_.has_value(), "Response must have http data to be converted into http response");
+
+    http::response<http::string_body> result{httpData_->status, httpData_->version};
+    result.set(http::field::server, fmt::format("clio-server-{}", util::build::getClioVersionString()));
+    result.set(http::field::content_type, asString(httpData_->contentType));
+    result.keep_alive(httpData_->keepAlive);
+    result.body() = std::move(message_);
+    result.prepare_payload();
+    return result;
+}
+
+boost::asio::const_buffer
+Response::asConstBuffer() const&
+{
+    ASSERT(not httpData_.has_value(), "Losing existing http data");
```

Comment:
Do we must check here ? Why message() does not need this check?

Reply:
`message()` just returns underlying message. `asConstBuffer()` is only used to send it by a websocket connection, so this method is only called inside websocket connection processing loop.